Faster dot product

mundya opened the pull request:
Uses the ARM `smlal` instruction to improve performance of neuron update step.
- [x] Fit model from IJCNN paper
  - [x] Update partitioning to account for this change in cost

These graphs are comparable to those in the IJCNN paper: the left hand bars show current master, the right hand bars show for this PR.

![profile_16_dimensions](https://cloud.githubusercontent.com/assets/3439837/9300901/63ef1608-44c2-11e5-81d2-be6610b62b6c.png)
![profile_100_neurons](https://cloud.githubusercontent.com/assets/3439837/9300902/64010c1e-44c2-11e5-8764-da2fa70b71f8.png)

mundya committed:
Optimised dot product

Introduces use of SMLAL for vector dot product in the neuron model.

mundya commented:
After 9725565: (bars in order: master, after a06676e, after 9725565)
![profile_16_dimensions](https://cloud.githubusercontent.com/assets/3439837/9327083/48c0be20-4596-11e5-8ef1-82373c730966.png)
![profile_100_neurons](https://cloud.githubusercontent.com/assets/3439837/9327085/4bc5d6c8-4596-11e5-9c32-2b6869dee940.png)

neworderofjamie commented:
Hah, take that less than operator :-/

On 18 August 2015 at 10:46, Andrew Mundy notifications@github.com wrote:

> After 9725565
> https://github.com/project-rig/nengo_spinnaker/commit/97255658ddefac31597c6d6875f1b981b5267407:
> (bars in order: master, after a06676e
> https://github.com/project-rig/nengo_spinnaker/commit/a06676e6469f442b09e13319cf25388c650785cc,
> after 9725565
> https://github.com/project-rig/nengo_spinnaker/commit/97255658ddefac31597c6d6875f1b981b5267407
> )
> [image: profile_16_dimensions]
> https://cloud.githubusercontent.com/assets/3439837/9327083/48c0be20-4596-11e5-8ef1-82373c730966.png
> [image: profile_100_neurons]
> https://cloud.githubusercontent.com/assets/3439837/9327085/4bc5d6c8-4596-11e5-9c32-2b6869dee940.png
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/project-rig/nengo_spinnaker/pull/58#issuecomment-132146632
> .

mundya committed:
Reduce spike processing cost

Reduces the cost of processing a spike.

mundya commented:
At this point we're here vs master (left hand bar is master, right hand is now).

![profile_16_dimensions](https://cloud.githubusercontent.com/assets/3439837/9334319/2324f46c-45c6-11e5-9d69-8b017f04dc24.png)
![profile_100_neurons](https://cloud.githubusercontent.com/assets/3439837/9334321/2328351e-45c6-11e5-9c6c-d6b638da05b7.png)

I slightly changed to model I was profiling against to give a better indication of reasonable firing rates.  That said there's still some space for someone to look into the distribution of neuron update costs.

![update_costs](https://cloud.githubusercontent.com/assets/3439837/9334455/d3a5d1e4-45c6-11e5-9c28-8b8207d4e891.png)

mundya commented:
For reference, profiling model:

``` python
with nengo.Network() as model:
    # Create the input to the network.
    ins = [nengo.Node(WhiteNoise(), label="inp") for _ in range(n_dims)]

    # Create the ensemble which we'll be profiling
    a = nengo.Ensemble(n_neurons, n_dims, label="A")

    # Create an ensemble to act as the end of the communication channel
    b = nengo.Ensemble(n_neurons, n_dims, label="B")

    # Connect the input to the first ensemble, and the first ensemble to
    # the second.
    for i, inp in enumerate(ins):
        nengo.Connection(inp, a[i], transform=1.0/np.sqrt(n_dims))
    nengo.Connection(a, b)

# Specify that we want to generate the input on chip and that we want to
# profile one of the ensembles.
nengo_spinnaker.add_spinnaker_params(model.config)
for node in model.all_nodes:
    model.config[node].function_of_time = True
model.config[a].profile = True
```

mundya commented:
OK, profiling!

| Cost | IJCNN Paper | Master | This PR |
| --- | --- | --- | --- |
| Input filtering | 43D + 245 | 42D + 181 | 40D + 196 |
| Neuron update | 13ND + 69N + 188 | 15ND + 88N | 9ND + 69N |
| Transmit output | 702D + 100 | 123D + 46 | 125D + 45 |

Where N is the number of neurons and D is the number of dimensions.
- _Input filtering_ Has changed little, as expected.
- _Neuron update_
  - I've removed the constant term as it wasn't significant
  - Between Master and This PR the neuron update cost decreases significantly
  - Between IJCNN and this PR the neuron model was improved and voltage recording added
- _Transmit output_ between IJCNN and Master I removed a 1us delay between transmitting each packet, hence the huge reduction in compute time.

mundya commented:
As a side note: that's around 2500 neurons per core in 1D or (the sadly unreachable due to memory boundedness) 925 neurons per core in 16D.

mundya committed:
Update neuron compute cost estimation.

Based on https://github.com/project-rig/nengo_spinnaker/pull/58#issuecomment-135969271

tcstewar commented:
> 925 neurons per core in 16D.

Awesome.  :) :)

neworderofjamie commented on spinnaker_components/common/fixed_point.h:
```diff
@@ -0,0 +1,97 @@
+/*****************************************************************************/
+// Convert from an INT64 to appropriate INT32 fixed point representation
+
+static inline int32_t scale_64_to_32(int64_t value)
```
This is a bit ambiguous  - convert_s32_30_s16_15 is also gross, but maybe more descriptive..

neworderofjamie commented on spinnaker_components/common/fixed_point.h:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  // Return the fixed point value
+  return result32;
```
What happens if you just shift down the 64-bit value and cast?

neworderofjamie commented on spinnaker_components/ensemble/ensemble_update.c:
```diff
@@ -77,23 +78,45 @@ void ensemble_update(uint ticks, uint arg1) {
-      // Set the voltage to be the overshoot, set the refractory time
-      set_neuron_refractory(n);
-      set_neuron_voltage(n, v_voltage - 1.0k);
+    if (bitsk(v_voltage) <= bitsk(1.0k))
```
Add comment as to the horror

mundya committed:
Merge branch 'master' into intrinsics

Conflicts:
	nengo_spinnaker/binaries/nengo_ensemble.aplx
	nengo_spinnaker/binaries/nengo_ensemble_profiled.aplx
	spinnaker_components/ensemble/ensemble.h
	spinnaker_components/ensemble/ensemble_update.c

mundya committed:
Rename `scale_64_to_32`

Changed to `convert_s32_30_s16_15` as per review by @neworderofjamie

mundya committed:
Use shift to simplify code/ASM

mundya committed:
Make use of bitsk in comparisons clear

mundya commented:
All done.

mundya commented:
Merging after checking with @neworderofjamie.
